File: src/main/java/com/ptaf/ai/triage/TriageService.java
package com.ptaf.ai.triage;

import com.ptaf.ai.GeminiClient;
import com.ptaf.ai.audit.AiGenerationAuditLogger;
import com.ptaf.ai.audit.AiGenerationAuditRecord;
import com.ptaf.ai.config.AiAssistantProperties;
import com.ptaf.ai.policy.AiPolicy;

import java.nio.file.Path;
import java.time.Instant;
import java.util.List;
import java.util.UUID;

/**
 * Phase 4: sends failure logs / stack traces to Gemini for a short structured triage note.
 */
public final class TriageService {

    private final GeminiClient client = new GeminiClient();

    public String triage(String logExcerpt, AiAssistantProperties props) throws Exception {
        return triage(logExcerpt, props, new AiPolicy());
    }

    public String triage(String logExcerpt, AiAssistantProperties props, AiPolicy policy) throws Exception {
        String safe = policy.maybeRedactTriageInput(logExcerpt != null ? logExcerpt : "");
        String system = """
                You are a test automation triage assistant for Java, Cucumber, Playwright, and Maven.
                Analyze the log excerpt. Be concise and practical.

                Respond with EXACTLY these markers:
                <<<LIKELY_AREA>>>
                (one of: UI | API | DB | CONFIG | ENV | FLAKY | UNKNOWN)
                <<<END_LIKELY_AREA>>>
                <<<SUMMARY>>>
                2-4 sentences: what failed and the most probable cause.
                <<<END_SUMMARY>>>
                <<<SUGGESTED_FILES>>>
                - optional bullet lines: relative paths in the repo to inspect (guess from stack traces)
                <<<END_SUGGESTED_FILES>>>
                """.strip();

        String user = "LOG EXCERPT:\n" + safe;
        String out = client.generateRaw(system, user, props);
        AiGenerationAuditRecord auditRecord = new AiGenerationAuditRecord(
                UUID.randomUUID().toString(),
                Instant.now().toString(),
                "triage",
                "service",
                props.model(),
                props.promptVersion(),
                AiGenerationAuditLogger.sha256(safe),
                "n/a",
                true,
                true,
                true,
                false,
                0,
                0,
                0,
                List.of(),
                List.of()
        );
        new AiGenerationAuditLogger().append(
                Path.of(System.getProperty("user.dir")),
                props.auditEnabled(),
                props.auditOutputPath(),
                auditRecord
        );
        return out;
    }
}

File: python/ptaf/ai/triage/triage_service.py
from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path
from uuid import uuid4

from ptaf.ai.audit.ai_generation_audit_logger import AiGenerationAuditLogger
from ptaf.ai.audit.ai_generation_audit_record import AiGenerationAuditRecord
from ptaf.ai.config.ai_assistant_properties import AiAssistantProperties
from ptaf.ai.gemini_client import GeminiClient
from ptaf.ai.policy.ai_policy import AiPolicy


class TriageService:
    def __init__(self, client: GeminiClient | None = None) -> None:
        self._client = client or GeminiClient()

    def triage(
        self,
        log_excerpt: str,
        props: AiAssistantProperties,
        policy: AiPolicy | None = None,
    ) -> str:
        active_policy = policy or AiPolicy()
        safe = active_policy.maybe_redact_triage_input(log_excerpt or "") or ""
        system = (
            "You are a test automation triage assistant for Python, pytest-bdd, Playwright, and uv.\n"
            "Analyze the log excerpt. Be concise and practical.\n\n"
            "Respond with EXACTLY these markers:\n"
            "<<<LIKELY_AREA>>>\n"
            "(one of: UI | API | DB | CONFIG | ENV | FLAKY | UNKNOWN)\n"
            "<<<END_LIKELY_AREA>>>\n"
            "<<<SUMMARY>>>\n"
            "2-4 sentences: what failed and the most probable cause.\n"
            "<<<END_SUMMARY>>>\n"
            "<<<SUGGESTED_FILES>>>\n"
            "- optional bullet lines: relative paths in the repo to inspect (guess from stack traces)\n"
            "<<<END_SUGGESTED_FILES>>>"
        )
        user = f"LOG EXCERPT:\n{safe}"
        out = self._client.generate_raw(system, user, props)
        audit_record = AiGenerationAuditRecord(
            request_id=str(uuid4()),
            timestamp=datetime.now(timezone.utc).isoformat(),
            command="triage",
            mode="service",
            model=props.model(),
            prompt_version=props.prompt_version(),
            requirement_hash=AiGenerationAuditLogger.sha256(safe),
            output_path="n/a",
            parse_successful=True,
            step_validation_passed=True,
            yaml_validation_passed=True,
            file_written=False,
            reused_steps_count=0,
            new_steps_count=0,
            missing_yaml_keys_count=0,
            warnings=[],
            blocking_errors=[],
        )
        AiGenerationAuditLogger().append(
            Path.cwd(),
            props.audit_enabled(),
            props.audit_output_path(),
            audit_record,
        )
        return out
